File: desktop/openmic/protocol.py
"""Wire protocol shared between the mobile app and the desktop server.

Single UDP socket, one byte packet type prefix:
  HELLO       -> announces a phone connecting, payload is the device name (UTF-8)
  AUDIO_PCM   -> sequence number (4 bytes, big-endian) + raw PCM16 mono samples
  AUDIO_OPUS  -> sequence number (4 bytes, big-endian) + Opus-encoded frame
  BYE         -> phone disconnecting, no payload
  PAIR_CHAL   -> desktop -> phone: pairing challenge with PIN
  PAIR_RESP   -> phone -> desktop: user confirmed PIN
  PAIR_ACK    -> desktop -> phone: pairing confirmed, includes device_id + token

v1 HELLO format: [0x01, 0x01, device_name]
v1 HELLO (paired): [0x01, 0x01, device_id (16 bytes) + auth_token (16 bytes) + device_name]
"""

import secrets
import struct

HELLO = 0x01
AUDIO = 0x02      # raw PCM16 (legacy)
BYE = 0x03
PAIR_CHAL = 0x04
PAIR_RESP = 0x05
PAIR_ACK = 0x06
AUDIO_OPUS = 0x07  # Opus-encoded audio frame

PROTOCOL_VERSION = 0x01

SAMPLE_RATE = 48000
CHANNELS = 1
SAMPLE_WIDTH = 2  # 16-bit PCM

# Opus encoding settings
OPUS_BITRATE = 24000          # 24 kbps - voice quality, ~10x smaller than PCM
OPUS_FRAME_SIZE_MS = 20       # 20ms frames
OPUS_FRAME_SAMPLES = SAMPLE_RATE * OPUS_FRAME_SIZE_MS // 1000  # 960 samples per frame

# Selectable quality presets. Only the encoder (phone) needs to know the
# bitrate — an Opus frame carries its own configuration, so the desktop
# decoder handles any of these without being told which one is in use.
OPUS_BITRATE_MIN = 12000      # intelligible speech on a congested network
OPUS_BITRATE_MAX = 64000      # near-transparent voice, ~2.7x the default
OPUS_BITRATE_PRESETS = (16000, 24000, 48000)


def clamp_bitrate(value: int) -> int:
    """Keep a requested bitrate inside the range libopus handles for voice."""
    return max(OPUS_BITRATE_MIN, min(OPUS_BITRATE_MAX, int(value)))

DEVICE_ID_LEN = 16
AUTH_TOKEN_LEN = 16
PIN_LEN = 6  # digits

_AUDIO_HEADER = struct.Struct(">I")


def pack_hello(device_name: str) -> bytes:
    return bytes([HELLO, PROTOCOL_VERSION]) + device_name.encode("utf-8")


def pack_hello_paired(device_id: bytes, auth_token: bytes, device_name: str) -> bytes:
    """HELLO for already-paired device: version + device_id + auth_token + name."""
    return (
        bytes([HELLO, PROTOCOL_VERSION])
        + device_id
        + auth_token
        + device_name.encode("utf-8")
    )


def pack_hello_v0(device_name: str) -> bytes:
    """Pack a pre-versioning HELLO for pairing challenge (no version byte)."""
    return bytes([HELLO]) + device_name.encode("utf-8")


def pack_audio(sequence: int, pcm: bytes) -> bytes:
    return bytes([AUDIO]) + _AUDIO_HEADER.pack(sequence) + pcm


def pack_audio_opus(sequence: int, opus_data: bytes) -> bytes:
    return bytes([AUDIO_OPUS]) + _AUDIO_HEADER.pack(sequence) + opus_data


def pack_bye() -> bytes:
    return bytes([BYE])


def pack_pair_challenge(pin: str) -> bytes:
    """Desktop -> Phone: pairing challenge with PIN."""
    return bytes([PAIR_CHAL]) + pin.encode("ascii")


def pack_pair_response(pin: str) -> bytes:
    """Phone -> Desktop: user confirmed PIN, echoing it back for validation."""
    return bytes([PAIR_RESP]) + pin.encode("ascii")


def pack_pair_ack(device_id: bytes, auth_token: bytes) -> bytes:
    """Desktop -> Phone: pairing confirmed, includes device_id + auth_token."""
    return bytes([PAIR_ACK]) + device_id + auth_token


def generate_device_id() -> bytes:
    return secrets.token_bytes(DEVICE_ID_LEN)


def generate_auth_token() -> bytes:
    return secrets.token_bytes(AUTH_TOKEN_LEN)


def generate_pin() -> str:
    return "".join(str(secrets.randbelow(10)) for _ in range(PIN_LEN))


def unpack(packet: bytes):
    if not packet:
        raise ValueError("empty packet")
    packet_type = packet[0]
    if packet_type == HELLO:
        if len(packet) < 3:
            raise ValueError("HELLO packet too short for version+name")
        version = packet[1]
        if version == PROTOCOL_VERSION:
            # Check if paired HELLO (has device_id + auth_token = 32 bytes before name)
            if len(packet) >= 2 + DEVICE_ID_LEN + AUTH_TOKEN_LEN:
                device_id = packet[2 : 2 + DEVICE_ID_LEN]
                auth_token = packet[2 + DEVICE_ID_LEN : 2 + DEVICE_ID_LEN + AUTH_TOKEN_LEN]
                name = packet[2 + DEVICE_ID_LEN + AUTH_TOKEN_LEN :].decode("utf-8", errors="replace")
                return HELLO, (version, device_id, auth_token, name)
            # Regular v1 HELLO
            return HELLO, (version, None, None, packet[2:].decode("utf-8", errors="replace"))
        # Backward compat: treat as pre-versioning HELLO
        return HELLO, (0, None, None, packet[1:].decode("utf-8", errors="replace"))
    if packet_type == AUDIO:
        (sequence,) = _AUDIO_HEADER.unpack(packet[1:5])
        return AUDIO, (sequence, packet[5:])
    if packet_type == AUDIO_OPUS:
        (sequence,) = _AUDIO_HEADER.unpack(packet[1:5])
        return AUDIO_OPUS, (sequence, packet[5:])
    if packet_type == BYE:
        return BYE, None
    if packet_type == PAIR_CHAL:
        pin = packet[1:].decode("ascii", errors="replace")
        return PAIR_CHAL, pin
    if packet_type == PAIR_RESP:
        # v1: echoes the PIN as an ASCII payload for validation; v0: empty.
        pin = packet[1:].decode("ascii", errors="replace")
        return PAIR_RESP, pin
    if packet_type == PAIR_ACK:
        if len(packet) < 1 + DEVICE_ID_LEN + AUTH_TOKEN_LEN:
            raise ValueError("PAIR_ACK too short")
        device_id = packet[1 : 1 + DEVICE_ID_LEN]
        auth_token = packet[1 + DEVICE_ID_LEN : 1 + DEVICE_ID_LEN + AUTH_TOKEN_LEN]
        return PAIR_ACK, (device_id, auth_token)
    raise ValueError(f"unknown packet type: {packet_type}")

File: mobile/lib/protocol.dart
import 'dart:convert';
import 'dart:typed_data';

/// Wire protocol shared with the desktop app. See desktop/openmic/protocol.py
/// and docs/protocol.md for the full spec.
class Protocol {
  static const int hello = 0x01;
  static const int audio = 0x02;        // raw PCM16 (legacy)
  static const int bye = 0x03;
  static const int pairChal = 0x04;
  static const int pairResp = 0x05;
  static const int pairAck = 0x06;
  static const int audioOpus = 0x07;    // Opus-encoded audio frame

  static const int version = 0x01;

  static const int sampleRate = 48000;
  static const int channels = 1;

  // Opus settings (must match desktop)
  static const int opusBitrate = 24000;
  static const int opusFrameSizeMs = 20;
  static const int opusFrameSamples = 960;  // 48000 * 20 / 1000

  // Selectable quality presets. Only the encoder needs to know the bitrate:
  // an Opus frame carries its own configuration, so the desktop decodes any
  // of these without being told which one is in use.
  static const int opusBitrateMin = 12000;
  static const int opusBitrateMax = 64000;
  static const List<int> opusBitratePresets = <int>[16000, 24000, 48000];

  /// Keep a requested bitrate inside the range libopus handles for voice.
  static int clampBitrate(int value) {
    if (value < opusBitrateMin) return opusBitrateMin;
    if (value > opusBitrateMax) return opusBitrateMax;
    return value;
  }

  /// Output-buffer size that caps one frame at [bitrate].
  ///
  /// opus_dart 3.x exposes no OPUS_SET_BITRATE control, but libopus treats
  /// the output buffer it is handed as a hard ceiling for that frame and
  /// lowers quality to fit — so sizing the buffer per frame is how the
  /// quality presets are actually enforced. It is a ceiling, not a target:
  /// quiet audio still encodes smaller.
  static int maxFrameBytesFor(int bitrate) {
    final framesPerSecond = 1000 ~/ opusFrameSizeMs;
    return clampBitrate(bitrate) ~/ 8 ~/ framesPerSecond;
  }

  static const int deviceIdLen = 16;
  static const int authTokenLen = 16;
  static const int pinLen = 6;

  static Uint8List packHello(String deviceName) {
    final nameBytes = utf8.encode(deviceName);
    final packet = Uint8List(2 + nameBytes.length);
    packet[0] = hello;
    packet[1] = version;
    packet.setRange(2, packet.length, nameBytes);
    return packet;
  }

  static Uint8List packHelloPaired(List<int> deviceId, List<int> authToken, String deviceName) {
    final nameBytes = utf8.encode(deviceName);
    final packet = Uint8List(2 + deviceIdLen + authTokenLen + nameBytes.length);
    packet[0] = hello;
    packet[1] = version;
    packet.setRange(2, 2 + deviceIdLen, deviceId);
    packet.setRange(2 + deviceIdLen, 2 + deviceIdLen + authTokenLen, authToken);
    packet.setRange(2 + deviceIdLen + authTokenLen, packet.length, nameBytes);
    return packet;
  }

  /// Pack a pre-versioning HELLO for pairing challenge (no version byte).
  static Uint8List packHelloV0(String deviceName) {
    final nameBytes = utf8.encode(deviceName);
    final packet = Uint8List(1 + nameBytes.length);
    packet[0] = hello;
    packet.setRange(1, packet.length, nameBytes);
    return packet;
  }

  static Uint8List packAudio(int sequence, Uint8List pcm) {
    final packet = Uint8List(5 + pcm.length);
    packet[0] = audio;
    packet.buffer.asByteData().setUint32(1, sequence, Endian.big);
    packet.setRange(5, packet.length, pcm);
    return packet;
  }

  static Uint8List packAudioOpus(int sequence, Uint8List opusData) {
    final packet = Uint8List(5 + opusData.length);
    packet[0] = audioOpus;
    packet.buffer.asByteData().setUint32(1, sequence, Endian.big);
    packet.setRange(5, packet.length, opusData);
    return packet;
  }

  static Uint8List packBye() {
    return Uint8List.fromList([bye]);
  }

  static Uint8List packPairResponse(String pin) {
    final bytes = utf8.encode(pin);
    final packet = Uint8List(1 + bytes.length);
    packet[0] = pairResp;
    packet.setRange(1, packet.length, bytes);
    return packet;
  }

  /// Returns (packetType, payload) or null if invalid.
  /// Payload varies by type:
  /// - HELLO: {version: int, deviceId: Uint8List?, authToken: Uint8List?, name: String}
  /// - PAIR_CHAL: String (PIN)
  /// - PAIR_ACK: {deviceId: Uint8List, authToken: Uint8List}
  /// - AUDIO: (sequence, pcm) — not parsed here, handled directly
  /// - PAIR_RESP, BYE: null
  static dynamic unpack(Uint8List packet) {
    if (packet.isEmpty) return null;
    final type = packet[0];
    switch (type) {
      case hello:
        // v1 protocol uses PROTOCOL_VERSION byte; v0 (pre-versioning) has name at packet[1].
        if (packet.length < 3) return null;
        final version = packet[1];
        if (version == Protocol.version) {
          // Check if paired HELLO (has device_id + auth_token = 32 bytes before name)
          if (packet.length >= 2 + deviceIdLen + authTokenLen) {
            final deviceId = packet.sublist(2, 2 + deviceIdLen);
            final authToken = packet.sublist(2 + deviceIdLen, 2 + deviceIdLen + authTokenLen);
            final name = utf8.decode(packet.sublist(2 + deviceIdLen + authTokenLen));
            return {
              'type': hello,
              'version': version,
              'deviceId': deviceId,
              'authToken': authToken,
              'name': name,
            };
          }
          // Regular v1 HELLO
          return {
            'type': hello,
            'version': version,
            'deviceId': null,
            'authToken': null,
            'name': utf8.decode(packet.sublist(2)),
          };
        }
        // Backward compat: treat as pre-versioning HELLO
        return {
          'type': hello,
          'version': 0,
          'deviceId': null,
          'authToken': null,
          'name': utf8.decode(packet.sublist(1)),
        };
      case pairChal:
        if (packet.length < 1 + pinLen) return null;
        return {
          'type': pairChal,
          'pin': utf8.decode(packet.sublist(1, 1 + pinLen)),
        };
      case pairAck:
        if (packet.length < 1 + deviceIdLen + authTokenLen) return null;
        return {
          'type': pairAck,
          'deviceId': packet.sublist(1, 1 + deviceIdLen),
          'authToken': packet.sublist(1 + deviceIdLen, 1 + deviceIdLen + authTokenLen),
        };
      case bye:
        return {'type': bye};
      default:
        return null;
    }
  }
}
